Doc/regions.md

```diff
@@ -0,0 +1,453 @@
+#  it points at a fresh dictionary object, which is also local.
+y = {}
+#  y is a field of the frame object, which is local. 
+#  it points at a fresh dictionary object, which is also local.
+x.f = y
+# x.f is a field of the dictionary object, which is local.
+# y is also local so no need to do anything in the write barrier.
+
+# Create a fresh region
+r = Region()
+# r is an explicit region
+r.f = y
+# r.f is a field of the region object, which is explicit.
+# y was local, so we need to promote y to the explicit region.
+# Before the assignment y has an RC of 2: One from the frame object, one from the object pointed to be x.
+# This adds two local RCs to the explicit region.
+y = None
+# This reduces the localRC on r, by one. 
+x = None
+# This causes the old value of x to be collected
+# This in turn reduces the localRC on r by one as it clears the f field that points into the explicit region.
+
+# We can now send the region r. It only has a single local RC r, and the operation will cause the 
+# local frame to clear that reference.
+send(r)
+# r is None at this point
+```
+
+To re-establish the correct local reference count, we calculate the updates to LRC as follows.
+First we calculate the set of local objects, L, that are being made reachable from this explicit region.
+As the set L is only reachable by local objects, we know that its reference count is purely from local objects.  
+Thus, we calculate sum of all the rcs in L, called R.
+Some of these references are internal to the set L,
+so we calculate the number of references from L to L, called I.
+Thus, we have created R-I additional references into the region.
+However, L can also reach into the explicit region, so we calculate the number of references from L to the explicit region, C.
+Thus, we need to increase the explicit region's local reference count by R - I, and then subtract C from it.
+This re-establishes the invariant.
+
+
+### Immutable objects
+
+To share objects between behaviours, we need to ensure that the object is not mutated.
+This is achieved by making the object "deeply immutable".
+This means that the object and all objects reachable from it are also immutable.
+
+The types in Python are mutable, but are typically only mutated at the start of the program.
+To support this, we need an operation to turn initially mutable objects into immutable objects.
+We add to the programming model a `freeze` operation that makes an object and its reachable graph of objects immutable.
+
+The fields of an immutable object may not be modified.
+Its reference count can be modified and it can be collected if it is no longer referenced.
+This reference count may be concurrently modified.
+
+To allow globals to be used in a concurrent setting, we have an explicit operation to promote make the globals immutable.
+```
+make_globals_immutable()
+```
+This must have been run to start the concurrency runtime and enable explicit region creation.
```

Comment:
If that is the case, why not just call it during the initialisation of the `veronapy` module when we do the other startup things, e.g. starting the subinterpreters.

Reply:
Because we need some time to allow for the set up of the types and any monkey patching.

As Types are globals, you wouldn't be able to have any fields on Types (for example default values) if they are immutable from the start.